File: app/src/main/java/com/jeppsson/appexplore/FeaturesUtils.java
package com.jeppsson.appexplore;

import android.content.pm.FeatureInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import android.support.annotation.RequiresApi;

class FeaturesUtils {
    static String getFeatures(PackageInfo packageInfo) {
        StringBuilder sb = new StringBuilder();

        if (packageInfo.reqFeatures != null) {
            for (FeatureInfo featureInfo : packageInfo.reqFeatures) {
                if (featureInfo.name == null) {
                    sb.append(reqGlEsVersionToString(featureInfo));
                } else {
                    sb.append(featureInfo.name)
                            .append(" (")
                            .append(featureFlagsToString(featureInfo));
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                        sb.append(featureVersionToString(featureInfo));
                    }
                    sb.append(')');
                }
                sb.append('\n');
            }
        }

        return sb.toString().trim();
    }

    private static String featureFlagsToString(FeatureInfo featureInfo) {
        return "required="
                + ((featureInfo.flags & FeatureInfo.FLAG_REQUIRED) == 0 ? "false" : true);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private static String featureVersionToString(FeatureInfo featureInfo) {
        return featureInfo.version > 0 ? ", version=" + featureInfo.version : "";
    }

    private static String reqGlEsVersionToString(FeatureInfo featureInfo) {
        return featureInfo.name == null ? "reqGlEsVersion=" + featureInfo.getGlEsVersion() : "";
    }

    public static String getFeaturesNoAvailable(PackageManager packageManager, PackageInfo packageInfo) {
        StringBuilder sb = new StringBuilder();

        if (packageInfo.reqFeatures != null) {
            for (FeatureInfo featureInfo : packageInfo.reqFeatures) {
                if (featureInfo.name != null) {
                    if (!packageManager.hasSystemFeature(featureInfo.name)) {
                        sb.append(featureInfo.name)
                                .append(" (")
                                .append(featureFlagsToString(featureInfo));
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                            sb.append(featureVersionToString(featureInfo));
                        }
                        sb.append(')');
                    }
                }
                sb.append('\n');
            }
        }

        return sb.toString().trim();
    }
}

File: app/src/main/java/com/jeppsson/appexplore/utils/FeaturesUtils.kt
package com.jeppsson.appexplore.utils

import android.content.pm.FeatureInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Build
import androidx.annotation.RequiresApi

internal object FeaturesUtils {
    fun getFeatures(packageInfo: PackageInfo): String {
        val sb = StringBuilder()

        if (packageInfo.reqFeatures != null) {
            for (featureInfo in packageInfo.reqFeatures) {
                if (featureInfo.name == null) {
                    sb.append(reqGlEsVersionToString(featureInfo))
                } else {
                    sb.append(featureInfo.name)
                            .append(" (")
                            .append(featureFlagsToString(featureInfo))
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                        sb.append(featureVersionToString(featureInfo))
                    }
                    sb.append(')')
                }
                sb.append('\n')
            }
        }

        return sb.toString().trim { it <= ' ' }
    }

    private fun featureFlagsToString(featureInfo: FeatureInfo): String {
        return "required=" + (featureInfo.flags and FeatureInfo.FLAG_REQUIRED != 0)
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private fun featureVersionToString(featureInfo: FeatureInfo): String {
        return if (featureInfo.version > 0) ", version=" + featureInfo.version else ""
    }

    private fun reqGlEsVersionToString(featureInfo: FeatureInfo): String {
        return if (featureInfo.name == null) "reqGlEsVersion=" + featureInfo.glEsVersion else ""
    }

    fun getFeaturesNoAvailable(packageManager: PackageManager, packageInfo: PackageInfo): String {
        val sb = StringBuilder()

        if (packageInfo.reqFeatures != null) {
            for (featureInfo in packageInfo.reqFeatures) {
                if (featureInfo.name != null) {
                    if (!packageManager.hasSystemFeature(featureInfo.name)) {
                        sb.append(featureInfo.name)
                                .append(" (")
                                .append(featureFlagsToString(featureInfo))
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                            sb.append(featureVersionToString(featureInfo))
                        }
                        sb.append(')')
                    }
                }
                sb.append('\n')
            }
        }

        return sb.toString().trim { it <= ' ' }
    }
}
